### ISSUE TITLE
Light artefacts on rough MeshStandardMaterial with IBL setup

### ISSUE THREAD

**Issue by silvainSayduck** [mrdoob/three.js#16409]
##### Description of the problem

I updated from r100 to r104, and somewhere along the way, something went wrong with IBL setup: fully rough materials have light artefacts (and that even in parts which are occluded by occlusion maps). It's especially visible on dark materials.
Because of this, it is impossible to get the feeling of dark fabric for example, since the material always has some reflection.

See this fully black, fully rough product comparison. The only light in the scene is an EXR envMap used for IBL. This also happens when using a JPG envMap for IBL, so I'm thinking it might have something to do with the CubeMap generator...

r104
![](https://i.imgur.com/PQEK2MK.png)

r100
![](https://i.imgur.com/ZL6nr7S.png)

And here is a minimal reproduction of the issue, where light artefacts are visible on the torus mesh, making it look like it has reflections while it should be fully rough: https://jsfiddle.net/4er2nxwt/

##### Three.js version

- [ ] Dev
- [x] r104

**Comment by WestLangley** [comment]
Experimenting with your example fiddle, it appears this change occurred in r.101.

A reasonable guess is the visual change occurred with https://github.com/mrdoob/three.js/pull/15644.

/ping @jsantell 

Your model may benefit from an ambient occlusion map to darken leaking light, but that is a separate issue.

**Comment by silvainSayduck** [comment]
Hi @WestLangley 

Thanks a lot for the quick answer, and good to know something indeed went wrong!

[As for my model's AO map, it is in fact using one (which makes this issue even more problematic, as these light artefacts seem not impacted - or not sufficiently - by AO maps). But as you said, that is a different issue]

**Comment by silvainSayduck** [comment]
FYI, another side effect is that rough metals now stay white, instead of being gray.
See screenshots from https://threejs.org/examples/webgl_materials_envmaps_exr.html

Roughness 1, metalness 1
![](https://i.imgur.com/OomhmMh.png)

Roughness 1, metalness 0
![](https://i.imgur.com/HsVs91g.png)

**Comment by jsantell** [comment]
Can you try reverting this PR and see if it fixes it? https://github.com/mrdoob/three.js/pull/15673

I've suspected that the above fix brightened the fresnel to match r100 but reverting it gives correct energy preservation for both dielectric and conductors when taking the whole material into account; I hope to look into this idea soon, but would like more opinions on the reverted change in different environments (and if it improves in this case)

**Comment by silvainSayduck** [comment]
I believe this fiddle uses a version of three r104, with that specific PR #15673 reverted, but it doesn't seem to fix much, unless I just did it wrong:

http://jsfiddle.net/0atr9v62/
http://jsfiddle.net/0atr9v62/1/

**Comment by jsantell** [comment]
I believe https://github.com/mrdoob/three.js/pull/16559 landing should fix this issue, @silvainSayduck can you give r105 a try?

> FYI, another side effect is that rough metals now stay white, instead of being gray.

This is intended/desired IMO, if I understand correctly, such that the color/energy is consistent across roughness

**Comment by mrdoob** [comment]
The minimal reproduction in the first post seems to be fully rough now 🙌

https://jsfiddle.net/4er2nxwt/

**Comment by silvainSayduck** [comment]
Hi everyone,

Just gave the r105 a try, and although the issue is solved for dark rough materials, now there seems to be the opposite problem, with dark edges appearing on light glossy materials...

![](https://i.imgur.com/HNR7DMI.png)
https://jsfiddle.net/mabr2xuq/ (r105)

---
![](https://i.imgur.com/AY7cbUn.png)
http://jsfiddle.net/qaf3zL0k/ (r104)

**Comment by jsantell** [comment]
The darkened fresnel in that case can be explained here https://github.com/mrdoob/three.js/pull/16559#issuecomment-496022231

I believe it's correctly rendering the specular light via darkened areas in the environment map, resulting in that effect, noticeable [in this example](https://threejs.org/examples/webgl_materials_envmaps_hdr.html) when you rotate the camera to reflect the sky rather than the darker areas. With that perspective, maybe the fresnel's not receiving the (white in this case) diffuse

**Comment by silvainSayduck** [comment]
Ok! Thanks for your answer @jsantel :) I see that the reversion has been merged into dev, so I'm assuming r106 should end up fixing that dark fresnel too.

Thanks again!

**Comment by silvainSayduck** [comment]
Hi again,

It seems that r106 still has the darkened fresnel (https://jsfiddle.net/mabr2xuq/). Did I understand wrong and this updated rendering has been added to three.js for good?

In my humble opinion, the very visible dark outlines look quite unrealistic: it's for example impossible to render a simple white ball, as it gets strong gray outlines which clearly does not happen in real life...

**Comment by jsantell** [comment]
> Hi again,
> 
> It seems that r106 still has the darkened fresnel (https://jsfiddle.net/mabr2xuq/). Did I understand wrong and this updated rendering has been added to three.js for good?
> 
> In my humble opinion, the very visible dark outlines look quite unrealistic: it's for example impossible to render a simple white ball, as it gets strong gray outlines which clearly does not happen in real life...

@silvainSayduck  What metal/roughness values are you referring to, what is a simple white ball? The edge isn't visible on lighter environments, e.g. this "edge" is the color of the map. The low resolution of the mesh exacerbates the effect in the link I think. AFAICT it's correct mathwise and correctly reflecting the environment and passing furnace tests. What does the same scene in Babylon or [Khronos' sample viewer](https://github.com/KhronosGroup/glTF-Sample-Viewer) or Filament look like? It's still difficult to evaluate PBR outside of a reference, and different trade-offs along the way make even that a challenge.

TLDR: To the extent of my knowledge, this is correct rendering, and of course I very well could be wrong on math and rendering expectations -- if this is incorrect rendering, I think comparing it to another PBR system would be the best way of confirming that

**Comment by silvainSayduck** [comment]
Hi @jsantell ,

Thanks a lot for your quick an detailed answer. Here is what I meant:

r104 https://jsfiddle.net/ev6uo4hw/
![](https://i.imgur.com/1kRTHSB.png)

--

r106 https://jsfiddle.net/ev6uo4hw/1/
![](https://i.imgur.com/c45EgHG.png)

--

Babylon https://www.babylonjs-playground.com/#MITBWZ
![](https://i.imgur.com/L86uWpW.png)

--

And as I said, I believe that having darkened edges is just not how objects would behave in the real world...

**Comment by jsantell** [comment]
@silvainSayduck this is a great example, thank you! It looks as if while the edges are dependent on the map's color (only occurs when the dark portions of the map are on the edge), it seems exacerbated for smooth dielectrics (metal: 0, roughness: 0), with exposure really highlighting the erroneously dark edges. I'll take a look at this over the next few days, but I don't have any initial ideas as what's going on other than looking at the fresnel more

**Comment by silvainSayduck** [comment]
Hi @jsantell,

Have you had time to do some more experiments and tests around this? Thanks a lot again :)

**Comment by silvainSayduck** [comment]
Hi again @mrdoob and @jsantell,

Just to check again on this issue: is there any way I could maybe help with this if @jsantell does not have time? Or has it now been decided that the dark edges are expected behavior?

Thanks in advance!

**Comment by WestLangley** [comment]
@silvainSayduck 

This image from my testbed (using your environment map) shows clearly that the blurring in PMREM is [buggy](https://github.com/mrdoob/three.js/pull/15390#issuecomment-445425354).

![Screen Shot 2019-11-01 at 11 15 39 AM](https://user-images.githubusercontent.com/1000017/68035244-a205dd80-fc99-11e9-8424-58b7adeb79e5.png)

PMREM is going to be replaced by @elalish -- hopefully very soon, but I do not know exactly when.

Your issue is important, but I think we should pursue this after PMREM is replaced.

**Comment by silvainSayduck** [comment]
Hi! I saw that the PMREM was replaced, amazing work!

I also just took a moment to check ![the example again on r112 with the new PMREM](https://threejs.org/examples/webgl_materials_envmaps_exr.html), which seems to still show the dark outline issue on glossy light materials. What would then be the next step regarding this?

Thanks a lot again for the great work and massive improvements :)

**Comment by elalish** [comment]
@silvainSayduck Thank you! This thread has gotten pretty long and wandered from issue to issue, so maybe the best thing would be to start a new issue, especially with up-to-date sphere renders, as those have been the most clear at expressing your problem.

**Comment by WestLangley** [comment]
@Slayvin Using the new PMREM, I see this:

![Screen Shot 2020-01-23 at 1 48 12 PM](https://user-images.githubusercontent.com/1000017/73014169-b31d4100-3de7-11ea-8236-734908a1cce1.png)

The background in this image is your environment map. It appears to me that the black edge is partially due the black regions of your map.

//

The three.js IBL multi-scattering code is based on [this paper](http://www.jcgt.org/published/0008/01/03/).

If you read Section 4, the author addresses the issue you raise, and argues that the darker edges are an improvement.

That does not mean there is not a three.js bug, but the "furnace tests" at least behave as one would expect.

If you feel there is something incorrect, you can help by opening a new issue and providing a live example that can be compared to other rendering engines.

**Comment by silvainSayduck** [comment]
Hi!

Thanks a lot for the quick reaction and the pointers! I will investigate this again, and come back with a clean new issue as per your suggestions afterwards if I believe it is still needed.

Thanks again for the help!
